### DISCUSSION TITLE
Use dedicated SMuFL glyphs for changing clefs

### DISCUSSION THREAD

**Discussion by rettinghaus** [rism-digital/verovio#2852] [Ideas]
This discussion is meant as a follow-up to #2849.

Classic engravings usually use different relative sizes for different changing clefs. 
Here's an example from the _Neue Liszt Ausgabe_ with a changing G clef at 80% size and a changing F clef slightly larger at 82% size.

![image](https://user-images.githubusercontent.com/7693447/167630764-fbc695e8-f645-46a8-ad07-7cae316644de.png)

Using the pre-defined SMuFL glyphs for changing clefs could improve stylistic consistency.
I took the liberty to make a slight adjustment in Leipzig font and put together an example with the dedicated glyphs in Verovio: 

### Leipzig

![changing_clefs_Leipzig](https://user-images.githubusercontent.com/7693447/167626328-28ba3d1e-0627-488a-aa4e-c93f81a1a43b.png)
(G clef 76.5%, F clef 81.4%)

### Leland

![changing_clefs_Leland](https://user-images.githubusercontent.com/7693447/167628529-468bf0ce-24fe-4b0e-b2ab-8e8d63bc2607.png)

### Bravura

![changing_clefs_Bravura](https://user-images.githubusercontent.com/7693447/167628574-f6235033-32d5-4c6a-ab9f-ee3122e95aff.png)

Only problem I see now is that we currently scale the margins of smaller clefs accordingly to the defined factor, so maybe that needs to be adjusted.

**Comment by craigsapp** [comment]
A ha &mdash; so maybe I was complaining about Bravura's change-clef size (after switching to it due to the tremolo beam problem in Leland :-).   But in verovio 3.2 this size was larger? (but the comparison tool is probably using Leipzig, so Leipzig changed size after version 3.2?)  Did Bravura shrink the clef or did verovio change from shrinking the regular clef to get a change-clef or did it switch to using the clef-change glyph for clef changes?   This also explains why Dorico is using a 66% size clef change, since it is using the Bravura font.

In any case there are two problems with that treble clef:

<img width="91" alt="Screen Shot 2022-05-10 at 6 06 54 AM" src="https://user-images.githubusercontent.com/3487289/167635431-75e01bda-6b20-4f00-8549-45213f86f422.png">

(1) It is distracting that the top loop does not reach the top staff line.   This makes it look like a cue clef change.
(2) The circular terminator on the clef is touching the bottom staff line.

The Liszt example demonstrates that the clef-change glyphs should perhaps not be rescaled to a uniform size within verovio, since it is actually a different glyph and not a scaled-down version of the regular clef.

--------------------------

The Leland font in verovio is also small (all at 66%):

<img width="504" alt="Screen Shot 2022-05-10 at 6 35 03 AM" src="https://user-images.githubusercontent.com/3487289/167641096-a4ad8f59-1a02-4cdc-9771-66a2085d9baa.png">

Is that due to the font, or to verovio scaling the clef changes?

Here is what the clef changes look like in SCORE (which the Leland font is based on):

<img width="722" alt="Screen Shot 2022-05-10 at 6 33 15 AM" src="https://user-images.githubusercontent.com/3487289/167641376-a6740cd1-47ac-4464-a4e4-2b7909a1ff91.png">

Here the bass clef change is better and closer to the modern standard than the Liszt example: There is a gap between the top staff and the clef, while in the Liszt there is none (so it is less apparent that the clef is a clef change since only the bottom of the clef is different.   In any case, a clef change that is too large (80-82%) is better than one that is too small (66%).  The Liszt bass clef change is also not uniform: The horizontal scaling is the same for the entire glyph, but the vertical scaling is different for each half of the glyph: above the F line, there is no scaling and below the F line there is some scaling.

------------------------------------

I see something different for Leipzig than you do:

<img width="1017" alt="Screen Shot 2022-05-10 at 6 55 38 AM" src="https://user-images.githubusercontent.com/3487289/167646293-f0ec7895-1d55-4ee6-b2d4-bb1fe49b2e1c.png">

So you are looking directly at the font (76.5% scaling)?  And I am looking at it after some extra scaling that verovio adds (66.7%)?

It seems like verovio is normalzing the clef-change sizes, but the problem is that it is normalizing it to 66% rather than 75% (and I don't care if it is exactly 75%, but that is a more reasonable default sicne 66% look unnatural).

**Comment by rettinghaus** [comment]
Right now Verovio scales every clef just down to the value given in `--clef-change-factor`, and this is what you see. What I put together is a version where Verovio uses the special [SMuFL glyphs](https://w3c.github.io/smufl/latest/tables/clefs.html) for changing clefs (U+E07A–U+E07C). 
And the corresponding glyphs in Leipzig have been specially prepared by me for this purpose (with the given rations) and could be adjusted even further. 
(Sorry that this wasn't clear in the first place.)

**Comment by craigsapp** [comment]
Thinking about it some more: 

(1) Verovio should use the clef-change glyph from a font without altering the size by default (using the size of the clef-change as it is expressed in the font).   That scaling could be different in different fonts, and at the decision of the font designer. I think this was true in version 3.2 of verovio, and changed afterwards.

(2) Verovio users could change the default size of clef change by using `--clef-change-factor` to change to a new size.  This would be easiest to implement by referring to the 100% size of the regular clef (such that a regular clef is 100%).  Since clef change sizes can be different in different fonts, verovio should scale according to the size of the regular clef, but instead scale the clef-change glyph of the font, so the actual scaling of the clef change glyph would be in reference to the nominal regular clef size rather than the nominal  change clef size (i.e., shrinking a regular clef is not optimal since the widths of thin lines should not be scaled).

This would mean that there is no default value for `--clef-change-factor` since the default would be to use the clef-change glyph without any additional scaling.

The word "factor" in `--clef-change-factor` is a bit ambiguous (but not a real problem).  Another possibility for the option name would be `--clef-change-size` or `--clef-change-scale`.

**Comment by rettinghaus** [comment]
Fonts updated and implemented in #2885.
